gt/communication/ghex_comm_mt.hpp

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include "./ghex_comm.hpp"
+#include "./factorize.hpp"
+#include <ghex/glue/gridtools/field.hpp>
+
+namespace communication {
+
+namespace ghex_comm {
+
+
+class grid_mt
+{
+public: // member types
+  using domain_id_type    = local_domain::domain_id_type;
+  using transport_type    = ::gridtools::ghex::tl::mpi_tag;
+  using grid_type         = ::gridtools::ghex::structured::grid::template type<local_domain>;
+  using patterns_type     = ::gridtools::ghex::pattern_container<transport_type, grid_type, domain_id_type>;
+  using patterns_ptr_t    = std::unique_ptr<patterns_type>;
+
+protected: // member types
+    using coordinate_type = local_domain::coordinate_type;
+    using domain_vec      = std::vector<local_domain>;
+    using comm_obj_type   = typename std::remove_reference<typename std::remove_cv<decltype(
+                               ::gridtools::ghex::make_communication_object<patterns_type>())>::type>::type;
+    using comm_obj_ptr_t  = std::unique_ptr<comm_obj_type>;
+
+public:
+    struct sub_grid
+    {
+        domain_id_type      domain_id;
+        MPI_Comm            mpi_comm;
+        comm_obj_ptr_t      co;
+        patterns_type*      patterns;
+        vec<std::size_t, 2> global_resolution;
+        vec<std::size_t, 2> offset;
+        vec<std::size_t, 3> resolution;
+    };
+
+protected: // members
+    int                 m_size;
+    std::array<int, 2>  m_dims;
+    MPI_Comm            m_comm_cart;
+    int                 m_rank;
+    std::array<int, 2>  m_coords;
+    coordinate_type     m_first;
+    coordinate_type     m_last;
+    halo_generator      m_hg;
+    vec<std::size_t, 2> m_global_resolution;
+    domain_vec          m_domains;
+    patterns_ptr_t      m_patterns;
+    moved_bit           m_moved;
+
+public: // ctors
+    grid_mt(vec<std::size_t, 3> const &global_resolution, int num_sub_domains = 1)
+    : m_size{[this]()
```

Comment:
These arguments could be const, but still, I would prefer having all this code just in the constructor body